### ISSUE TITLE
Types from `internal::*` should not appear in public types

### ISSUE THREAD

**Issue by briansmith** [rustls/rustls#920]
In error.rs we have:
```rust
enum Error {
    ...
    /// We received a fatal alert.  This means the peer is unhappy.
    AlertReceived(AlertDescription),
}
```

`AlertDescription` seems to be only reachable by importing `rustls::internal::msgs::AlertDescription`.

The description for `rustls::internal` says:
```
/// Internal classes which may be useful outside the library.
/// The contents of this section DO NOT form part of the stable interface.
```

However, our intent is that the entirety of `Error` is part of the stable API.

I propose we re-export `pub use msgs::AlertDescription` within errors.rs so that stable users can avoid depending on the unstable `rustls::internal` API.

This comes up when one wants to write tests that verify that a certain `AlertReceived` error is returned by Rustls.

**Comment by jbr** [comment]
`AlertDescription` was made pub from the root in 5dab6ed654a35d8c3af2468732cb3410ab1dc32b, and the crate currently builds without internal being pub. Should this issue become "conditionally expose `rustls::internal` behind a default-disabled feature flag?" in order to prevent this sort of thing in the future?

**Comment by djc** [comment]
Yes, that sounds like a good improvement.

**Comment by jbr** [comment]
are we ok adding a dependency on (visibility)[https://crates.io/crates/visibility] in order to make functions conditionally `pub`/`pub(crate)` to remove unreachable_pub lints?

**Comment by djc** [comment]
One example here is the `Codec` trait. This is faux-public through internals, and as such shows up in the docs for any types that implement it -- many of which are public (`AlertDescription` is one example). Ultimately if someone wants to use the trait they'll have to import it and the only way of doing so is through `internals` so maybe it's not much worse than the other things, but probably still something to consider.

**Comment by jsha** [comment]
We also have this problem with NamedGroup, Tls12ClientSessionValue, and Tls13ClientSessionValue in the new ClientSessionStore trait. In the documentation they link to rustls::internal::msgs::enums::NamedGroup, rustls::internal::msgs::persist::Tls12ClientSessionValue, and rustls::internal::msgs::persist::Tls13ClientSessionValue.

/cc #1224

**Comment by jsha** [comment]
Here's the current list of `internals` types that are reachable in the public interface, borrowing a search command from https://github.com/rustls/rustls/issues/1036#issuecomment-1140389210:

```
base::PayloadU16
codec::Reader
handshake::DigitallySignedStruct
handshake::DistinguishedNames
handshake::SessionID
handshake::VecU16OfPayloadU16
persist::ClientSessionCommon
persist::Tls12ClientSessionValue
persist::Tls13ClientSessionValue
```

These are the ones that are new since 0.20.8:

```
handshake::SessionID
persist::ClientSessionCommon
persist::Tls12ClientSessionValue
persist::Tls13ClientSessionValue
```

**Comment by ctz** [comment]
> persist::Tls12ClientSessionValue
> persist::Tls13ClientSessionValue

These are exported as `rustls::client::Tls12ClientSessionValue` and `rustls::client::Tls13ClientSessionValue`: https://github.com/rustls/rustls/blob/main/rustls/src/lib.rs#L424-L426

**Comment by jsha** [comment]
> handshake::SessionID

This is no longer part of the public API thanks to #1226. 👍🏻

> These are exported as rustls::client::Tls12ClientSessionValue and rustls::client::Tls13ClientSessionValue: https://github.com/rustls/rustls/blob/main/rustls/src/lib.rs#L424-L426

Unfortunately, rustdoc still chooses to link to the `internal::` path from ClientSessionStore. And I believe rust-analyzer may also autoimport the `internal::` variant.

In https://github.com/rustls/rustls/issues/1036#issuecomment-1124747345 @djc suggested that once an `internal` item becomes part of the public API, we should move it out of the `msgs` directory so it's clear that it's public. That also happens to neatly fix the problem of an item existing with under two paths. I think that would be a good approach for TLS1[23]ClientSessionValue.

> persist::ClientSessionCommon

This is public via Tls1[23]ClientSessionValue, since it's a public member. I assume we want to treat it as fully part of the public API?

**Comment by djc** [comment]
I think it would probably make sense to move all the stuff from `enums` along with the `KeyExchangeAlgorithm` enum from `handshake`. I think we should definitely not make `DigitallySignedStruct`, `DistinguishedNames` and `VecU16OfPayloadU16` public in their current form, and probably avoid doing so for `PayloadU16` as well.

Was this latest list from current main too, so that `SessionID` is still public even after #1226? Is there an easy way of figuring out how (via what path) each of these becomes public?

**Comment by jsha** [comment]
> Was this latest list from current main too, so that SessionID is still public even after https://github.com/rustls/rustls/pull/1226?

Yep:

![image](https://user-images.githubusercontent.com/220205/226729188-e0c84bd7-58de-4836-be1b-1150dd66c316.png)

> Is there an easy way of figuring out how (via what path) each of these becomes public?

```
$ rg SessionID target/doc/ -l -g '*.html' -g '!target/doc/rustls/internal/*' -g '!*.rs.html'
target/doc/rustls/all.html
target/doc/rustls/client/struct.Tls12ClientSessionValue.html
target/doc/rustls/msgs/handshake/struct.SessionID.html
```

**Comment by djc** [comment]
FWIW, you did not cover traits here, but we should probably consider those, too (`Codec` comes to mind).

**Comment by jsha** [comment]
Updated command that includes traits. Running this only turns up `Codec` as an additional output:

```
cargo doc --all-features && find target/doc/rustls -name '*.html' | grep -v internal/ | xargs sed -n 's/.*\(struct\|type\|enum\|trait\) rustls::internal::msgs::\([A-Za-z]\+::[A-Za-z0-9]\+\).*/\1 \2/p' | sort -u
```

What should we do about [ClientSessionCommon](https://rustdoc.crud.net/jsha/rustls-main/doc/rustls/internal/msgs/persist/struct.ClientSessionCommon.html)? Tls12ClientSessionValue [gets its methods by Deref'ing to it](https://rustdoc.crud.net/jsha/rustls-main/doc/rustls/client/struct.Tls12ClientSessionValue.html#deref-methods-ClientSessionCommon). Same for [Tls13ClientSessionValue](https://rustdoc.crud.net/jsha/rustls-main/doc/rustls/client/struct.Tls13ClientSessionValue.html#deref-methods-ClientSessionCommon).

I think our options are:

 - Publicize ClientSessionCommon. This helps users who want to implement both a client and a server session store using similar code.
  - Un-publicize ClientSessionCommon, and instead of Deref'ing to it, use public passthrough methods on Tls1[23]ClientSessionValue.

**Comment by cpu** [comment]
Is there a way to tighten the visibility of `ClientSessionCommon.rewind_epoch`: https://github.com/rustls/rustls/blob/d8f04295baa990b0d7168dfdac09c228a5dca2c5/rustls/src/msgs/persist.rs#L230-L234 (I think only used from the bogo shim).

I could be off track but if the type becomes public it would put that fn in front of more users and now seems like the time to change it if we can :crossed_fingers:

**Comment by jsha** [comment]
> It doesn't abstract over client vs server, only over client 1.2 vs 1.3, right?

Yep, misspoke! Thanks for the correction.

> I think the API becomes pretty crappy if we forego the Deref approach, so maybe publicizing is better?

What makes it crappy? I was thinking, e.g.:

```rust
impl Tls12ClientSessionValue {
  pub fn server_cert_chain(&self) -> &[key::Certificate] {
    self.common.server_cert_chain()
}
...
```

And so on. Then I realized that in #1230, all the methods on `ClientSessionCommon` became private. Are you saying the internal API becomes crappy without Deref?

As I currently understand it, there's nothing a user can do with a `ClientSessionCommon` other than Clone or Debug it.

**Comment by ctz** [comment]
yeah afaict the only reason `ClientSessionCommon` is pub is so bogo_shim can  call `rewind_epoch` on it, and the Deref doesn't work there because it needs mutation. how about instead: https://github.com/rustls/rustls/pull/1232

**Comment by djc** [comment]
Yes, the internal API becomes crappy. See for example the `Retrieved::<T>::map()` impl that I added earlier today.

> As I currently understand it, there's nothing a user can do with a `ClientSessionCommon` other than Clone or Debug it.

Well, they can technically `rewind_epoch()` on it. Should be clear that that's not something they're intended to do...

> Is there a way to tighten the visibility of `ClientSessionCommon.rewind_epoch`:

Not a straightforward way, no. Examples and integration tests count as external to the crate.

**Comment by cpu** [comment]
> yeah afaict the only reason ClientSessionCommon is pub is so bogo_shim can call rewind_epoch on it, and the Deref doesn't work there because it needs mutation. how about instead: https://github.com/rustls/rustls/pull/1232

:bulb: That's a nice way forward
